org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServerWrapper.java

```diff
@@ -387,11 +389,12 @@ private void advanceinitializeFutureMonitor() {
 		}
 	}
 
-	private synchronized Job createInitializeLanguageServerJob() {
+	private Job createInitializeLanguageServerJob() {
 		return new Job(NLS.bind(Messages.initializeLanguageServer_job, serverDefinition.label)) {
 			@Override
 			protected IStatus run(IProgressMonitor monitor) {
-				initializeFutureMonitor = SubMonitor.convert(monitor, initializeFutureNumberOfStages);
```

Comment:
Maybe it is a silly question, but do we need an `AtomicReference`? Could we not achieve the same by saving `initializeFutureMonitor` to a local variable in line 394 (and also on line 382) before saving it to the class field, and then mark the monitor as done using the local variable?

Is there a benefit of `AtomicReference` over the local variable?

Reply:
As far as I see it, at a minimum the initializeFutureMonitor needs to be made volatile because it is accessed from two different threads and the Job in createInitializeLanguageServerJob otherwise may never see the assigned value. The nice part of AtomicReference is that it offers compareAndSet() which can avoid another type of race conditions when the wrapper is stopped and restarted in quick succession.